pkg/common/jwtutil/keyset.go

```diff
@@ -0,0 +1,141 @@
+package jwtutil
+
+import (
+	"context"
+	"encoding/json"
+	"io"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"github.com/zeebo/errs"
+	"gopkg.in/square/go-jose.v2"
+)
+
+type KeySetProvider interface {
+	GetKeySet(context.Context) (*jose.JSONWebKeySet, error)
+}
+
+type KeySetProviderFunc func(context.Context) (*jose.JSONWebKeySet, error)
+
+func (fn KeySetProviderFunc) GetKeySet(ctx context.Context) (*jose.JSONWebKeySet, error) {
+	return fn(ctx)
+}
+
+type KeySetConfigURL string
+
+func (c KeySetConfigURL) GetKeySet(ctx context.Context) (*jose.JSONWebKeySet, error) {
+	uri, err := FetchKeySetURI(ctx, string(c))
+	if err != nil {
+		return nil, err
+	}
+	return FetchKeySet(ctx, uri)
+}
+
+type CachingKeySetProvider struct {
+	provider        KeySetProvider
+	refreshInterval time.Duration
+
+	mu      sync.Mutex
+	updated time.Time
+	jwks    *jose.JSONWebKeySet
+
+	hooks struct {
+		now func() time.Time
+	}
+}
+
+func NewCachingKeySetProvider(provider KeySetProvider, refreshInterval time.Duration) *CachingKeySetProvider {
+	c := &CachingKeySetProvider{
+		provider:        provider,
+		refreshInterval: refreshInterval,
+	}
+	c.hooks.now = time.Now
+	return c
+}
+
+func (c *CachingKeySetProvider) GetKeySet(ctx context.Context) (*jose.JSONWebKeySet, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	now := c.hooks.now()
+
+	if !c.updated.IsZero() && now.Sub(c.updated) < c.refreshInterval {
+		return c.jwks, nil
+	}
+
+	// refresh key set. if there is a failure, log and return the old set if
+	// available.
+	jwks, err := c.provider.GetKeySet(ctx)
+	if err == nil {
+		c.jwks = jwks
+		c.updated = now
+	} else {
+		logrus.Warnf("unable to refresh key set: %v", err)
+		if c.jwks == nil {
+			return nil, errs.Wrap(err)
+		}
+	}
+
+	return c.jwks, nil
+}
+
+func FetchKeySetURI(ctx context.Context, configURL string) (string, error) {
```

Comment:
another option here could be `func DiscoverKeySetURI(ctx context.Context, oidcIssuer string) (string, error)`

Reply:
:+1: